Add cognito IDP guide (#1796)
FusionAuth/fusionauth-site: site/docs/v1/tech/identity-providers/openid-connect/cognito.adoc

---
layout: doc
title: Amazon Cognito OpenID Connect
description: Login using Cognito as an OpenID Connect Identity Provider
navcategory: login-methods
---

:sectnumlevels: 0

== Configure OpenID Connect with Cognito

[NOTE]
====
Cognito users must have a public email address configured to link on email (See link:/docs/v1/tech/identity-providers#linking-strategies[Linking Strategies for more on this]). An approach that will work for all users is to link on username or create an anonymous link. Using these strategies ensures that the configured Cognito OIDC connection works for every user, no matter their Cognito privacy settings.
====

Once you have completed this configuration, you may enable an OpenID Connect "Login with Cognito" button for one or more FusionAuth Applications. See link:https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-client-apps.html[Cognito - Configuring a user pool app client] for an additional reference.

image::identity-providers/cognito/cognito-login-fusionauth-button.png[Login with Cognito, width=1200, role=bottom-cropped]

=== Register a Cognito User Pool

> If you're interested in connecting to Cognito, it is likely that you already have a user pool set up that you'd want to connect to. We're adding the steps to create a new user pool in this guide in the interest of completeness, or in case you would like to set up a test user pool.

Reviewer: Is there a document on the AWS side you can link to on setting up a user pool? That might be clearer and kept up to date.

- Open the created app client and record both the "Client ID" and "Client Secret" which can be revealed by toggling the "Show client secret" button.
- The user pool and app client are now created and the next step is adding a test user.
- Open the user pool and under the users tab click "Create user".

=== Configure a New FusionAuth OpenID Connect Identity Provider

There is no pre-configured connector for Cognito in FusionAuth. The generic "OpenID" can be used though, as Cognito supports the standard OpenId Connect protocols.

Navigate to your FusionAuth instance. Select "Settings" from the sidebar, and then "Identity Providers".

Select "Add OpenID Connect" from the "Add" dropdown at the top right of the page.

image::identity-providers/cognito/cognito-choose-openid.png[Create a new OpenID integration, width=1200, role=bottom-cropped]

- Provide a "Name", like "Cognito".
- Set "Client Id" to the "App Client Id" recorded when creating the app client on Cognito.
- Select "HTTP Basic Authentication" for the "Client Authentication" field.
- Set "Client secret" to the app client secret recorded when creating the app client on Cognito.

> To retrieve the values for the various "endpoints" fields, visit:

Reviewer: Could you not use `Discover endpoints`? If that didn't work, please say "uncheck 'Discover endpoints'" or similar.

```
https://cognito-idp.<REGION>.amazonaws.com/<USER_POOL_ID>/.well-known/openid-configuration
```

Replace `<REGION>` with the AWS region code that you created your Cognito user pool in. This can be found by selecting the region indicator at the top right of the menu bar and recording the region code displayed alongside the region location.

Replace `<USER_POOL_ID` with the Cognito user pool Id. You can find this by clicking on your user pool.

image::identity-providers/cognito/cognito-user-pool-id.png[User Pool Id in Cognito, width=1200, role=bottom-cropped]

> Visiting the URL with your values templated in should show you values similar to these:

Reviewer: ```suggestion
If you browse to the `openid-configuration` URL should show you a JSON document similar to this:
```

```json
{
  "authorization_endpoint": "https://fusionauth.auth.us-east-2.amazoncognito.com/oauth2/authorize",
  "id_token_signing_alg_values_supported": [
    "RS256"
  ],
  "issuer": "https://cognito-idp.us-east-2.amazonaws.com/us-east-2_DFiufNrjlT",
  "jwks_uri": "https://cognito-idp.us-east-2.amazonaws.com/us-east-2_DFiufNrjlT/.well-known/jwks.json",
  "response_types_supported": [
    "code",
    "token"
  ],
  "scopes_supported": [
    "openid",
    "email",
    "phone",
    "profile"
  ],
  "subject_types_supported": [
    "public"
  ],
  "token_endpoint": "https://fusionauth.auth.us-east-2.amazoncognito.com/oauth2/token",
  "token_endpoint_auth_methods_supported": [
    "client_secret_basic",
    "client_secret_post"
  ],
  "userinfo_endpoint": "https://fusionauth.auth.us-east-2.amazoncognito.com/oauth2/userInfo"
}
```

NOTE: Ensure that the Domain name set earlier in the app client configuration appears in these endpoint URLs. If it does not, wait for a while for the URLs to update. There may be a lag between setting the domain and the domain reflecting in the URLs.

> Copy the endpoints from this output into the corresponding fields in the FusionAuth config.

Reviewer: Again, did discover endpoints not work? I would think it should.

Set "Button Text" to "Login with Cognito". You can also upload a Cognito icon as the button icon if you wish.

> Set the "Scope" field to "openid". Choose "Link on email. Create the user if they do not exist" for the "Linking Strategy".

Reviewer: Maybe put the warning that is at the top here too (as an include file).

> Choose "No Lambda" for the "Reconcile Lambda field".

Reviewer: ```suggestion
Choose "No Lambda" for the "Reconcile Lambda field". If you wanted to examine or modify the response of the Cognito authentication event and modify the user based on that, you'd link:/docs/v1/tech/lambdas/[create a lambda] and assign it here.
```

> Now choose the applications you would like the Cognito sign-in to be available to.

Reviewer: ```suggestion
Now choose the applications for which you would like the Cognito sign-in to be available.
```
